File: lab3/cpp/test_gost.cpp
#include "test_gost.h"
#include "general_fun.h"

#include <iostream>
#include <vector>
#include <math.h>
#include <iomanip>
#include <algorithm>
#include <random>

using namespace std;

unsigned int generate_N_1(int t){
    uint stop = 0;
    if (t == 2) return 3;
    double q = generate_N_1(t/2);
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> random_num(0.0, 1.0);
    double e;
    int N;
    int u = 0;
    unsigned int p;
    while (1){
        start:
        e = random_num(gen);
        N = ceil(pow(2, t-1)/q) + ceil(pow(2, (t-1))*e/q);
        N%2 == 1? N++: N=N;
        u = 0;
        stop++;
        if (stop > 1000000){
            return 0;
        }
        while (1){
            p = (N+u)*q+1;
            if (p > pow(2, t)) goto start;
            if(num_exp(2, p-1, p) == 1 && num_exp(2, N+u, p) != 1) return p;
            u += 2;
        }
    }
}

void gost(int k, int s = 2){ //k - длина числа в битах
    unsigned long N;
    vector<unsigned long> list_N;
    for (int i = 0, n = 0, test; i < 10;){  //заполняем список числами N
        N = generate_N_1(k);
        if (N == 0){
            cout << "Больше чисел не нашлось, генерация остановленна" << endl;
            return;
        }
        if (count(list_N.begin(), list_N.end(), N) == 0){
            list_N.push_back(N);
            cout << N << " по вероятностному тесту - " << test_millera_rabina(N, s) << endl;
            i++;
        }
    }
}

File: lab3/go/test_gost.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

// generate_N_1 генерирует простое число по алгоритму ГОСТ (только для t = 9, 10, 13, 14, 21, 22, 25, 26)
func generate_N_1(t int) uint64 {
	q := math.Ceil(float64(t) / 2)
	rand.Seed(time.Now().UnixNano())
	e := rand.Float64()
	N := math.Ceil(math.Pow(2, float64(t-1))/q) + math.Ceil(math.Pow(2, float64(t-1))*e/q)
	if int(N)%2 == 0 {
		N++
	}

	for {
		e = rand.Float64()
		N = math.Ceil(math.Pow(2, float64(t-1))/q) + math.Ceil(math.Pow(2, float64(t-1))*e/q)
		if int(N)%2 == 1 {
			N++
		}

		u := 0
		for {
			p := uint64((float64(N)+float64(u))*q + 1)
			if float64(p) > math.Pow(2, float64(t)) {
				break
			}
			if num_exp(2, uint64(p-1), uint64(p)) == 1 && num_exp(2, uint64(float64(N)+float64(u)), uint64(p)) != 1 {
				return p
			}
			u += 2
		}
	}
}

// gost реализует основной алгоритм ГОСТ
func gost(k int, s int) {

	var N uint64
	listN := make(map[uint64]bool)

	for i := 0; i < 10; {
		N = generate_N_1(k)
		if !listN[N] {
			listN[N] = true
			fmt.Printf("%d по вероятностному тесту - %s\n", N, test_millera_rabina(N, s))
			i++
		}
	}
}
